src/matching.rs

```diff
@@ -284,6 +284,20 @@ impl<T> DeclarationBlock<T> {
     }
 }
 
+bitflags! {
+    #[doc = "Flags set on elements during the matching process."]
+    flags ElementFlags: u8 {
+        #[doc = "When a child is added or removed from this element, all the children must be"]
+        #[doc = "restyled, because they may match selectors such as :last-of-type."]
+        const HAS_SLOW_SELECTOR = 0x01,
+        #[doc = "When a child is added or removed from this element, any later children must be"]
+        #[doc = "restyled, because they may match selectors such as :nth-child."]
+        const HAS_SLOW_SELECTOR_LATER_SIBLINGS = 0x02,
```

Comment:
Well, this patch as a whole is necessary for correctness when using incremental layout (which *is* an optimization).

Using `HAS_SLOW_SELECTOR` for all of these cases would be sufficient for correctness, since it would always restyle all the siblings.  The other flags are special cases where we can get away with restyling fewer siblings.  So adding more special cases is purely an optimization.